methods/guest/src/main.rs

```diff
@@ -1,24 +1,58 @@
 use risc0_zkvm::guest::env;
 
-use consensus::beacon_state;
-use consensus::beacon_block::BeaconBlock;
-use consensus::beacon_block_header::BeaconBlockHeader;
+use alloy_primitives::FixedBytes;
+use consensus_common::{
+    beacon_block::BeaconBlock,
+    beacon_block_header::BeaconBlockHeader,
+    funcs::{process_block_header, ProcessBlockHeaderOutput},
+    proof::Proof,
+};
+use ssz_rs::prelude::*;
 
 fn main() {
     // let count_start = env::cycle_count();
 
+    let pre_state_root: FixedBytes<32> = env::read();
+
     let slot: u64 = env::read();
-    let latest_block_header: BeaconBlockHeader = env::read();
+    let slot_proof: Proof = env::read();
+    let latest_block_header_ssz: Vec<u8> = env::read();
+    let latest_block_header: BeaconBlockHeader =
+        BeaconBlockHeader::deserialize(&mut latest_block_header_ssz.as_slice()).unwrap();
+    let latest_block_header_proof: Proof = env::read();
+    // let validator_slashed: bool = env::read();
+    // let validator_slashed_proof: Proof = env::read();
+
     let proposer_index: u64 = env::read();
-    let block: BeaconBlock = env::read();
 
-    let new_block_header = beacon_state::process_block_header(
-            slot,
-            &latest_block_header,
-            proposer_index,
-            &block,
-        )
-        .unwrap();
+    let block_ssz: Vec<u8> = env::read();
+    let block = BeaconBlock::deserialize(&mut block_ssz.as_slice()).unwrap();
+
+    // Verify proofs
+    slot_proof.verify().unwrap();
+    assert_eq!(slot_proof.index, 34);
+    assert_eq!(slot_proof.witness, pre_state_root);
+    latest_block_header_proof.verify().unwrap();
+    assert_eq!(latest_block_header_proof.index, 36);
+    assert_eq!(latest_block_header_proof.witness, pre_state_root);
+    // validator_slashed_proof.verify().unwrap();
+    // TODO: assert_eq for generalized index
+    // assert_eq!(validator_slashed_proof.witness, pre_state_root);
+
+    let new_block_header = process_block_header(
+        slot,
+        &latest_block_header,
+        false, // validator_slashed, TODO: fix
+        proposer_index,
+        &block,
```

Comment:
This is still consuming untrusted data. Even though the code has already done `slot_proof.verify()` and `latest_block_header_proof.verify()` but they're not linked to `slot` and `latest_block_header_proof`. A misbehaving host could still pass in `slot_proof` and `latest_block_header_proof` that are not for `slot` and `latest_block_header` respectively.

So before passing into `process_block_header()`, either we need to extract the data from the `Proof` (does `*_proof.leaf` contain the data?), or the zkvm hashes the `slot` and `latest_block_header_proof` and compares against the proof.

Reply:
This is checked on lines 33 and 36 with generalized indexes. IIUC, the slot generalized index will always be 34 and latest block header will be 36. So checking indices is enough?

`ssz-rs` should be able to reason about the generalized indices and verify proofs at those indices: https://github.com/ralexstokes/ssz-rs/blob/ec3073e2273b4d0873fcb6df68ff4eff79588e92/README.md#merkle-proofs

Reply:
I think `slot_proof.verify()` is enough without the index check actually.

But again I still think `slot` and `slot_proof` have no relationship to each other. The host can pass in a `slot` that's totally unrelated to the provided `slot_proof`.

A way to confirm this is if we change the [code here](https://github.com/ReamLabs/consenzero/pull/4/files#diff-060a0095cd9c61458a24214c0d43ab27f667d7f8c8ac4c60e592885a51ef2111R89):

```rs
.write(&pre_state_ssz_rs.slot)
```

to:

```rs
.write(1234) // or any number
```

I think the guest will still let it pass? Which means the guest is not enforcing state integrity. Expected behavior should be that the guest should fail that it cannot verify that the provided data belongs to the state root.

Reply:
Ok, I get your point. You are right, we need to connect the `slot` and `sloot_proof` somehow. I will look into it.

But I still think we need to check the index. Otherwise someone can provide the proof for some other u64 for example instead of slot (in the same beacon block) and the proof will be validated successfully.